src/toxcore/tcp/connections/connections.rs

```diff
@@ -0,0 +1,319 @@
+/*! TCP connections handling
+
+TCP connectins resides between net_crypto and tcp_client.
+It serves to net_crypto using tcp_client.
+It provides reliable connection to net_crypto via multiple tcp relays to a friend.
+When a Tox client connects to a friend via tcp relay, normally 3 redundant connections are established.
+One connection is used for data send/recv, 2 others are backup.
+In toxcore maximum number of redundant connections is 6.
+TCP connection can get into sleep mode.
+Getting into sleep mode can occur when UDP connecion is established, because Tox prefer UDP than TCP.
+When established UDP connection is disabled, TCP connecions are awaken.
+
+Defintion of struct names.
+
+
+>   Connections has
+         set of ConnOfClient(HashMap)
+         set of Connection(HashMap)
+     ConnOfClient has
+         object of TcpClient processor
+         IP, port, PK : to save for sleeping status
+     Connection has
+         object of 3 to 6 ConnToRelay connections for redundancy
+     ConnToRelay has
+         id_of_client as a key to ConnOfClient hashmap
+         connection_id of Routing Response packet
+
+*/
+
+use parking_lot::RwLock;
+
+use std::sync::Arc;
+use std::time::Instant;
+use std::collections::HashMap;
+use std::net::{SocketAddr, IpAddr, Ipv4Addr};
+
+use futures::sync::mpsc;
+use futures::{future, Future, Sink, Stream};
+use failure::{Error, err_msg};
+use tokio_codec::Framed;
+use tokio::net::TcpStream;
+
+use toxcore::time::*;
+use toxcore::crypto_core::*;
+use toxcore::tcp::codec::*;
+use toxcore::tcp::handshake::make_client_handshake;
+use toxcore::tcp::packet::*;
+
+/// The amount of maximum connections for each friend.
+pub const MAX_FRIEND_TCP_CONNECTIONS: usize =  6;
+
+/// The amount of recommended connections for each friend.
+///   NOTE: Must be at most (MAX_FRIEND_TCP_CONNECTIONS / 2).
+pub const RECOMMENDED_FRIEND_TCP_CONNECTIONS: usize =  MAX_FRIEND_TCP_CONNECTIONS / 2;
+
+/// Status of connection to TcpRelay.
+#[derive(PartialEq, Clone)]
+pub enum ConnectionStatus {
+    /// Uninitialized state.
+    None,
+    /// Connection is registered internally.
+    Registered,
+    /// Connection is online.
+    Online,
+}
+
+/// Status of connection to a friend.
+#[derive(PartialEq, Clone)]
+pub enum ConnOfClientStatus {
```

Comment:
Since there is `ConnectionStatus` above it'd look prettier if this struct had a name like `ClientConnectionStatus`.